change api links to xref format

mairaw opened the pull request:
Based on PR #3457, I've decided to tackle our @ notation format for links for two reasons:
1 - we standardize our link format
2 - I hope we get broken links warnings in the build report if any of those are broken. We don't get warnings with the @ notation if you're using a bad uid

mairaw committed:
change api links to xref format

mairaw committed:
bug fixes

mairaw committed:
fixed broken links

mairaw committed:
fixed broken links

mairaw committed:
fix broken link

mairaw committed:
fixed broken link

mairaw committed:
fixed broken link

rpetrusha commented on docs/csharp/tour-of-csharp/classes-and-objects.md:
```diff
@@ -134,7 +134,7 @@ An *output parameter* is used for passing arguments by reference. It's similar t
 [!code-csharp[OutExample](../../../samples/snippets/csharp/tour/classes-and-objects/OutExample.cs#L3-L17)]
 
-A *parameter array* permits a variable number of arguments to be passed to a method. A parameter array is declared with the `params` modifier. Only the last parameter of a method can be a parameter array, and the type of a parameter array must be a single-dimensional array type. The Write and WriteLine methods of the `@System.Console` class are good examples of parameter array usage. They are dec ...
+A *parameter array* permits a variable number of arguments to be passed to a method. A parameter array is declared with the `params` modifier. Only the last parameter of a method can be a parameter array, and the type of a parameter array must be a single-dimensional array type. The Write and WriteLine methods of the <xref:System.Console?displayProperty=nameWithType> class are good examples of pa ...
```
When used with a type, `displayProperty=nameWithType` displays only the type name, not the namespace. Did you want the namespace here? That requires `displayProperty=fullName`.

rpetrusha commented on docs/csharp/tutorials/console-webapiclient.md:
```diff
@@ -466,12 +466,12 @@ public DateTime LastPush
 property accessor throws an exception.
 
-To use @System.Globalization.CultureInfo.InvariantCulture, you will need to add the @System.Globalization namespace to the `using` statements 
+To use <xref:System.Globalization.CultureInfo.InvariantCulture>, you will need to add the <xref:System.Globalization namespace to the `using` statements 
```
There's no `>` after Globalization.

BillWagner commented on docs/csharp/whats-new/csharp-7.md:
```diff
@@ -162,7 +162,7 @@ Using tuples in this way offers several advantages:
 * You save the work of authoring a `class` or a `struct` that defines the type returned. 
 * You do not need to create new type.
-* The language enhancements removes the need to call the <xref:System.Tuple.Create``1(``0)> methods.
+* The language enhancements removes the need to call the <xref:System.Tuple.Create``1(``0)>> methods.
```
This looks like an extra `>` Is that right?

rpetrusha commented on docs/standard/exceptions/exception-class-and-properties.md:
```diff
@@ -19,7 +19,7 @@ manager: "wpickett"
 # Exception class and properties
 
-The @System.Exception class is the base class from which exceptions inherit. For example, the @System.InvalidCastException class hierarchy is as follows:
+The <xref:System.Exception> class is the base class from which exceptions inherit. For example, the <xref:System.InvalidCastException class hierarchy is as follows:
```
no closing `>` after InvalidCastException

rpetrusha commented on docs/standard/exceptions/exception-class-and-properties.md:
```diff
@@ -28,20 +28,20 @@ Object
+Most of the classes that inherit from <xref:System.Exception> do not implement additional members or provide additional functionality; they simply inherit from <xref:System.Exception>. Therefore, the most important information for an exception can be found in the hierarchy of exception classes, the exception name, and the information contained in the exception.
 
-It is recommended to throw and catch only objects that derive from @System.Exception, but you can throw any object that derives from the @System.Object class as an exception. Note that not all languages support throwing and catching objects that do not derive from @System.Exception.
+It is recommended to throw and catch only objects that derive from <xref:System.Exception>, but you can throw any object that derives from the <xref:System.Object> class as an exception. Note that not all languages support throwing and catching objects that do not derive from <xref:System.Exception>.
```
nit unrelated to your PR: It is recommended to throw --> We recommend that you throw

BillWagner reviewed (approved):
Looks good @mairaw 

I spotted a few discrepancies that I wanted to point out. Once you check those, :shipit:

rpetrusha commented on docs/standard/exceptions/index.md:
```diff
@@ -30,7 +30,7 @@ Applications must be able to handle errors that occur during execution in a cons
 An exception is any error condition or unexpected behavior that is encountered by an executing program. Exceptions can be thrown because of a fault in your code or in code that you call (such as a shared library), unavailable operating system resources, unexpected conditions that the runtime encounters (such as code that cannot be verified), and so on. Your application can recover from some of th ...
 
-In .NET, an exception is an object that inherits from the [System.Exception](xref:System.Exception) class. An exception is thrown from an area of code where a problem has occurred. The exception is passed up the stack until the application handles it or the program terminates.
+In .NET, an exception is an object that inherits from the <xref:System.Exception?displayProperty=nameWithType> class. An exception is thrown from an area of code where a problem has occurred. The exception is passed up the stack until the application handles it or the program terminates.
```
displayProperty=fullName if you want the namespace name to appear

rpetrusha commented on docs/visual-basic/programming-guide/concepts/collections.md:
```diff
@@ -180,11 +180,11 @@ End Class
+-   <xref:System.Collections.Concurrent> classes  
   
--   @System.Collections classes  
+-   <xref:System.Collections classes>  
```
`>` after Collections, not classes. This doesn't appear in the broken links report, though.

mairaw replied:
Probably because I didn't close it properly. That's why I have a feature request for CATS to find those post publishing

rpetrusha commented on includes/migration-guide/retargeting/wf/new-ambiguous-dispatcherinvoke-overloads-could-result-different-behavior.md:
```diff
@@ -2,7 +2,7 @@
 |   |   |
 |---|---|
-|Details|The .NET Framework 4.5 adds new overloads to Dispatcher.Invoke that include a parameter of type @System.Action. When existing code is recompiled, compilers may resolve calls to Dispatcher.Invoke methods that have a @System.Delegate parameter as calls to Dispatcher.Invoke methods with an @System.Action parameter. If a call to a Dispatcher.Invoke overload with a @System.Delegate parameter  ...
+|Details|The .NET Framework 4.5 adds new overloads to Dispatcher.Invoke that include a parameter of type <xref:System.Action>. When existing code is recompiled, compilers may resolve calls to Dispatcher.Invoke methods that have a <xref:System.Delegate> parameter as calls to Dispatcher.Invoke methods with an <xref:System.Action> parameter. If a call to a Dispatcher.Invoke overload with a <xref:Sys ...
```
nit: Dispatcher.Invoke should be a link: \<xref:System.Windows.Threading.Dispatcher.Invoke%2A?displayProperty=nameWithType>
UnobservedTaskException should be displayProperty=nameWithType
DispatcherOperation.Result should also be displayProperty=nameWithType

rpetrusha commented on docs/csharp/language-reference/keywords/throw.md:
````diff
@@ -27,7 +27,7 @@ The syntax of `throw` is:
 throw [e]
 ```
-where `e` is an instance of a class derived from <xref:System.Exception?displayProperty=nameWithType>. The following example uses the `throw` statement to throw an @System.IndexOutOfRangeException if the argument passed to a method named `GetNumber` does not correspond to a valid index of an internal array.
+where `e` is an instance of a class derived from <xref:System.Exception?displayProperty=nameWithType>. The following example uses the `throw` statement to throw an <xref:System.IndexOutOfRangeException> if the argument passed to a method named `GetNumber` does not correspond to a valid index of an internal array.
````
fullName instead of nameWithType? nameWithType for a type displays only the type name, not the fully qualified type name.

rpetrusha commented on docs/csharp/tutorials/console-webapiclient.md:
```diff
@@ -233,21 +233,21 @@ using System.Runtime.Serialization.Json;
 
 Secondly, the `as` operator converts from the compile time type of `object` to `List<repo>`. 
-The declaration of @System.Runtime.Serialization.Json.DataContractJsonSerializer.ReadObject(System.IO.Stream) declares that it returns an object of type <xref:System.Object?displayProperty=nameWithType>. @System.Runtime.Serialization.Json.DataContractJsonSerializer.ReadObject(System.IO.Stream) will return the type you specified when you constructed it (`List<repo>` in
+The declaration of <xref:System.Runtime.Serialization.Json.DataContractJsonSerializer.ReadObject(System.IO.Stream)> declares that it returns an object of type <xref:System.Object?displayProperty=nameWithType>. <xref:System.Runtime.Serialization.Json.DataContractJsonSerializer.ReadObject(System.IO.Stream)> will return the type you specified when you constructed it (`List<repo>` in
```
nameWithType, or fullName?

mairaw replied:
I didn't change that link. Let's see if we can get the behavior fixed. Otherwise, we'll have to review all links that use this parameter 😢

rpetrusha reviewed (changes requested):
Generally, this looks really good, @mairaw. A few nits, and a couple of discoveries:
- xrefs without a closing tag are not considered links and don't show up in the links report; they are just rendered as text.
- displayProperty=nameWithType when used with a type displays only the type name, not the fully qualified type name. To get the fully qualified type name, you have to use displayProperty=fullName

mairaw commented on docs/csharp/language-reference/keywords/when.md:
```diff
@@ -31,7 +31,7 @@ catch ExceptionType [e] when (expr)
 where *expr* is an expression that evaluates to a Boolean value. If it returns `true`, the exception handler executes; if `false`, it does not. 
 
-The following example uses the `when` keyword to conditionally execute handlers for an @System.Net.HttpRequestException depending on the text of the exception message.
+The following example uses the `when` keyword to conditionally execute handlers for an <xref:System.Net.Http.HttpRequestException> depending on the text of the exception message.
```
Yes. I've fixed it once we started getting warnings of the broken links. 😄

mairaw committed:
feedback

mairaw commented:
Thanks for the thorough review @rpetrusha and @BillWagner. I believe I've addressed all your comments with the exception of the nameWithType. I'm really hoping we can get one query parameter that will behave the way we want instead of having to think what kind of link we have.

rpetrusha commented:
LGTM now, @mairaw. I'll leave this for you to merge.
